### ISSUE TITLE
Trading Error for "order_buy_market" and "order_sell_market"

### ISSUE THREAD

**Issue by cpasean** [jmfernandes/robin_stocks#401]
Encountered error messages since 2 weeks ago: (robin-stocks ver: **3.0.4**) 

>>> rs.order_buy_market('AAPL', 1)
**Error in request_post: Expecting value: line 1 column 1 (char 0)**

>>> rs.order_sell_market('AAPL', 1)
**{'non_field_errors': ['Your app version is missing important stock trading updates. You can still place orders on the web.']}**

I will be grateful to have your advice ASAP, please help.

**Comment by SingularityMan** [comment]
This is directly related to the orders.order() method. I am still figuring it out on my end but every method that includes placing orders always calls orders.order, which as of robin_stocks 3.0.4 includes a modified payload. 

The problem lies in the payload itself, but I am still figuring out how to modify the payload to provide the correct set of parameters.

**Comment by meldinman** [comment]
Hey, Check out the pull request and my post on this issue #379  discussion where this issue has been going on...

**Comment by sanathnair09** [comment]
So today I received the same error when trying to do an order_buy_market. I added breakpoints and took a look at where the error was being raised from and these were my findings. 

Error is raised from helper.py Line 353 (the except block)
```python
def request_post(url, payload=None, timeout=16, json=False, jsonify_data=True):
    data = None
    res = None
    try:
        if json:
            update_session('Content-Type', 'application/json')
            res = SESSION.post(url, json=payload, timeout=timeout)
            update_session(
                'Content-Type', 'application/x-www-form-urlencoded; charset=utf-8')
        else:
            res = SESSION.post(url, data=payload, timeout=timeout)
        data = res.json()
    except Exception as message:
        print("Error in request_post: {0}".format(message), file=get_output())
    # Either return response <200,401,etc.> or the data that is returned from requests.
    if jsonify_data:
        return(data)
    else:
        return(res)
```

Essentially what is happening is that when an order_buy_market or sell is called the server responds with an HTTP 500 error. (making a speculation here but I'd assume this issue really occurs whenever the `def order(symbol, quantity, side, limitPrice=None, stopPrice=None, account_number=None, timeInForce='gtc', extendedHours=False, jsonify=True)` order method is called)

And then when the line `data = res.json()` is called, a JSONDecodeError is raised because the formatting of the 500 response doesn't seem to follow the guideline. (I'm not well versed in JSON standards but it seems the field doc was causing the issues.)

![Screenshot 2023-06-22 at 8 47 19 AM](https://github.com/jmfernandes/robin_stocks/assets/55722236/3c09d2dc-8ef4-49d1-b5fd-2f7f7d0bd97c)

**Comment by henryzhangpku** [comment]
It seems RH changed their POST API params due to the recent extended hours update.  If you open the chrome console, send a order from web (100% working) and observe the actual payload of POST request sent,  you will know what I mean.
  
However , at this moment  I did not figure out yet which exact params/headers need to add for this robin_stocks to re-work , will run a few more tests next week to get it done.

**Comment by henryzhangpku** [comment]
Btw: options trading does not have this same problem on stocks.  So it really is for recent extended hours update on stocks.

**Comment by henryzhangpku** [comment]
Looks like we have had a solution to fix this already https://github.com/jmfernandes/robin_stocks/issues/379#issuecomment-1563054858

**Comment by henryzhangpku** [comment]
Here is the PR to fix this issue:  https://github.com/jmfernandes/robin_stocks/pull/403
I tested it is working today even for overnight trading. See screenshot : 
![image](https://github.com/jmfernandes/robin_stocks/assets/2763638/4651d761-1d2f-498c-92ff-cea35a89b97f)

**Comment by cpasean** [comment]
Would you elaborate on how you cleared the error and made the order by sharing your code more in detail? Please...

**Comment by henryzhangpku** [comment]
Sure , here is my PR where you can see all my code fix : https://github.com/jmfernandes/robin_stocks/pull/403

RHS api is really all about REST.  Open your chrome console ,  place a trade via web (100% working), and from console/Network you will see the POST request sent to RHS /orders url.  Comparing the actual working POST request/payload to the code , you will see what params have been removed/added  by RHS in their new API version/POST request.  From that you get the answer to get it working . 

> Would you elaborate on how you cleared the error and made the order by sharing your code more in detail? Please...

**Comment by cpasean** [comment]
> Sure , here is my PR where you can see all my code fix : #403
> 
> RHS api is really all about REST. Open your chrome console , place a trade via web (100% working), and from console/Network you will see the POST request sent to RHS /orders url. Comparing the actual working POST request/payload to the code , you will see what params have been removed/added by RHS in their new API version/POST request. From that you get the answer to get it working .
> 
> > Would you elaborate on how you cleared the error and made the order by sharing your code more in detail? Please...


I really appreciate your tips. I got Buy Order fixed, but still struggling with Sell-order: 

>> order_sell_market('BIIB', 1)
**{'non_field_errors': ['Invalid Good Til Canceled order.']}**

Would you help me with fixing this 'non field errors'?

**Comment by henryzhangpku** [comment]
My PR has both buy/sell working. Take a look at my PR.

H 
> > Sure , here is my PR where you can see all my code fix : #403
> 
> > 
> 
> > RHS api is really all about REST. Open your chrome console , place a trade via web (100% working), and from console/Network you will see the POST request sent to RHS /orders url. Comparing the actual working POST request/payload to the code , you will see what params have been removed/added by RHS in their new API version/POST request. From that you get the answer to get it working .
> 
> > 
> 
> > > Would you elaborate on how you cleared the error and made the order by sharing your code more in detail? Please...
> 
> 
> 
> 
> 
> I really appreciate your tips. I got Buy Order fixed, but still struggling with Sell-order: 
> 
> 
> 
> >> order_sell_market('BIIB', 1)
> 
> **{'non_field_errors': ['Invalid Good Til Canceled order.']}**
> 
> 
> 
> Would you help me with fixing this 'non field errors'?

**Comment by cpasean** [comment]
> My PR has both buy/sell working. Take a look at my PR.
> 
> H
> 
> > > Sure , here is my PR where you can see all my code fix : #403
> > 
> > 
> > > 
> > 
> > 
> > > RHS api is really all about REST. Open your chrome console , place a trade via web (100% working), and from console/Network you will see the POST request sent to RHS /orders url. Comparing the actual working POST request/payload to the code , you will see what params have been removed/added by RHS in their new API version/POST request. From that you get the answer to get it working .
> > 
> > 
> > > 
> > 
> > 
> > > > Would you elaborate on how you cleared the error and made the order by sharing your code more in detail? Please...
> > 
> > 
> > I really appreciate your tips. I got Buy Order fixed, but still struggling with Sell-order:
> > > > order_sell_market('BIIB', 1)
> > 
> > 
> > **{'non_field_errors': ['Invalid Good Til Canceled order.']}**
> > Would you help me with fixing this 'non field errors'?

I've finally fixed the issue with your help. I appreciate that!

**Comment by paraluke23** [comment]
Will the PR 403 get merged soon? I cannot run it and am unable to follow the PR to change my code. Thanks.

**Comment by emily111123** [comment]
@henryzhangpku 

thanks for the scripts, this code in my computer: "url = orders_url() always shows:  it's not defined",could you please help me with this ? thanks

**Comment by Adelantado** [comment]
robin-stocks 3.0.5 has been out for a few days. Hope it helps.

**Comment by paraluke23** [comment]
No, 3.0.5 doesn't fix the issue: Error in request_post: Expecting value: line 1 column 1 (char 0)

**Comment by henryzhangpku** [comment]
> @henryzhangpku 
> 
> 
> 
> thanks for the scripts, this code in my computer: "url = orders_url() always shows:  it's not defined",could you please help me with this ? thanks
> 
> 
> 
> 

Can you share more details of this error ? I can help you out.

**Comment by doctorcolossus** [comment]
@henryzhangpku, nice work, but you seem to have forgotten to update the docstring, which would keep people from having to come fish around here in the issues to find out how this works. Also, `market_hours` in snake-case seems inconsistent with the `extendedHours` argument in camel-case, but maybe it's too late to change now to avoid breaking peoples' code. Anyway, thanks for the fix!

**Comment by henryzhangpku** [comment]
> @henryzhangpku, nice work, but you seem to have forgotten to update the docstring, which would keep people from having to come fish around here in the issues to find out how this works. Also, `market_hours` in snake-case seems inconsistent with the `extendedHours` argument in camel-case, but maybe it's too late to change now to avoid breaking peoples' code. Anyway, thanks for the fix!
One thing you should note is this API changes (from RH side obviously) , so from time to time the parameters have to be updated . I had a few recent changes on my end as well , with that I have 24 hours market 100% working.
